### PR TITLE
Pivotal ID # 179922370: Modify Submission Metadata

### PR DESCRIPTION
**Pull request by jhoanmanuelms**
https://www.pivotaltracker.com/story/show/179922370

- Add a new parameter to allow selecting a preferred files source
- Alter the files source order based on the preferred files source selected
- Update the CLI operations

### CODE REVIEW

File: submission/submission-webapp/src/main/kotlin/ac/uk/ebi/biostd/submission/domain/helpers/SourceGenerator.kt [hunk 1]
```diff
@@ -20,50 +23,53 @@ class SourceGenerator(
...
+            addUserSources(submitter, owner, rootPath)
+            addSubmissionSources(submission)
+        }
+
+        return ComposedFileSource(submissionSources, preferredSource)
     }
 
-    private fun submissionSources(
-        owner: SecurityUser?,
+    private fun MutableList<FilesSource>.addUserSources(
```
Comment (Juan-EBI) [inline submission/submission-webapp/src/main/kotlin/ac/uk/ebi/biostd/submission/domain/helpers/SourceGenerator.kt:40]: I am not sure about this, IMO it does not look correct (semantically not sintaxtically)that a mutable list has a business operation `addUserSources` I mean is a list which usually has operation as `elementAt(index)` or `addElement`.

**Review by Juan-EBI — COMMENTED** [review]
Looking good please check comments, also there is something missing in this PR, when we use submission source we need to get file size and md5 from database and not calcualted it so processing of a submission with submission sources is super fast.  As we have all the information in the database we may introduce a new type of ExtFile and instead of checking submission folder we check database when searching by path or name.

File: commons/commons-model-extended-mapping/src/main/kotlin/ebi/ac/uk/io/sources/ComposedFileSource.kt [hunk 2]
```diff
@@ -1,13 +1,31 @@
...
+    private val sources: List<FilesSource>,
+    private val preferredOrigin: FileOrigin
+) : FilesSource {
+    override val filesOrigin: FileOrigin
+        get() = MIXED
 
-    override fun getFile(path: String, md5: String?): File? =
-        sources.firstNotNullOfOrNull { it.getFile(path, md5) }
+    override fun getExtFile(path: String, md5: String?, attributes: List<Attribute>): ExtFile? {
+        val sorted = sources.filter { it.isPreferred() } to sources.filterNot { it.isPreferred() }
```
[the remark this answers is not in the document]
Comment (jhoanmanuelms) [reply commons/commons-model-extended-mapping/src/main/kotlin/ebi/ac/uk/io/sources/ComposedFileSource.kt:16]: I thought that at first but I was afraid that could be a very naive strategy. However, it's true that's us who build the sources array so it should stay ordered

**Comment by jhoanmanuelms** [comment]
These changes will be on hold until FIRE is deployed which most probably will require yet another PR

**Pull request closed without merging**